src/coreclr/jit/gentree.cpp

```diff
@@ -24133,6 +24165,390 @@ GenTree* Compiler::gtNewSimdSumNode(var_types type, GenTree* op1, CorInfoType si
+
+            GenTree* value1;
+            GenTree* value2;
+
+            switch (useFlags)
+            {
+                case TernaryLogicUseFlags::A:
+                {
+                    // One operand. We swap it with C so it can be contained where possible
+                    value1 = op1;
+                    value2 = nullptr;
+
+                    std::swap(op1, op3);
+                    break;
+                }
+
+                case TernaryLogicUseFlags::B:
+                {
+                    // One operand. We swap it with C so it can be contained where possible
+                    value1 = op2;
+                    value2 = nullptr;
+
+                    std::swap(op2, op3);
+                    break;
+                }
+
+                case TernaryLogicUseFlags::C:
+                {
+                    // One operand. It is already in the ideal position
+                    value1 = op3;
+                    value2 = nullptr;
+                    break;
+                }
+
+                case TernaryLogicUseFlags::AB:
+                {
+                    // Two operands. We swap A with C so it can be contained where possible
+                    // and so we aren't RMW
+
+                    value1 = op1;
+                    value2 = op2;
+
+                    std::swap(op1, op3);
+                    break;
+                }
+
+                case TernaryLogicUseFlags::AC:
+                {
+                    // Two operands. We swap A with B so we aren't RMW
+
+                    value1 = op1;
+                    value2 = op3;
+
+                    std::swap(op1, op2);
+                    break;
+                }
+
+                case TernaryLogicUseFlags::BC:
+                {
+                    // Two operands. They are already in the ideal positions
+
+                    value1 = op2;
+                    value2 = op3;
+                    break;
+                }
+
+                case TernaryLogicUseFlags::None:
+                {
+                    // No operands.
+
+                    value1 = nullptr;
+                    value2 = nullptr;
+
+                    break;
+                }
+
+                default:
+                {
+                    unreached();
+                }
+            }
+
+            switch (info.oper1)
+            {
+                case TernaryLogicOperKind::Select:
+                {
+                    assert(info.oper1Use != TernaryLogicUseFlags::None);
+                    assert(value2 == nullptr);
+
+                    assert(info.oper2 == TernaryLogicOperKind::None);
+                    assert(info.oper2Use == TernaryLogicUseFlags::None);
+
+                    assert((control == static_cast<uint8_t>(0xF0)) || // A
+                           (control == static_cast<uint8_t>(0xCC)) || // B
+                           (control == static_cast<uint8_t>(0xAA)));  // C
+
+                    return value1;
```

Comment:
What about side effects of all the other operands?

Reply:
Ah, right.

Is there a better way than inserting a comma if `op1` has side effects?
```cpp
GenTree* zero = gtNewZeroConNode(type);

if ((op1->gtFlags & GTF_SIDE_EFFECT) != 0)
{
    op1 = gtUnusedValNode(op1);
    op1 = gtNewOperNode(GT_COMMA, type, op1, zero);
}
else
{
    op1 = zero;
}
```

Reply:
Something like that should work. Though generally I would question doing this kind of folding from the constructor, do we really expect it to kick in for IL we're importing?

Reply:
This is a general handler, it's not strictly import specific and will be used in other places when the constant folding support is added.

While we're probably not super likely to see it for some cases where better alternatives exist (such as `A`, `B`, `C`, `and`, `or`, `xor`) we will see it for cases where `vpternlog` is the best way to do it (`not`, `nand`, `nor`, `xnor`).

In those cases there are many ways the user could specify the "unused" operand (specifying a constant, duplicating a value, etc). So this logic handles "normalizing" it to be in the "best shape" for the JIT. In particular, we want `A` to be the unused node since that is the `RMW` operand otherwise. Likewise we want the unused operand value to be something cheap that won't CSE'd and `constant zero` is a good candidate for that.

Reply:
Fixed.